Bitcoin backend generalization

darosior opened the pull request:
This standardizes our requests for Bitcoin data as a set of commands plugins can register, and makes our `bitcoind` interaction through `bitcoin-cli` the default plugin, which registers all of these commands.

This makes it possible to not rely on a bitcoin-core backend, which can help for instance for mobile integration (cf #3484).

See #3354 for some design decisions, but here is a summary:
- Separated plugins can register different calls, which allows plugins specialization (credits to @cdecker)
- Uses JSON-RPC commands instead of hooks (see #3354).
- Needs 5 commands to be registered by our plugin(s):
  -  `getchaininfo` - called at startup to get infos about our network, and about the block chain.
  -  `getrawblockatheight` - get the block at that height in hex format, along with its id.
  -  `sendrawtransaction` - self explanatory
  -  `gettxout` - get the amount and the scriptPubkey of this txo.
  -  `getfeerate` - get the feerate in **btc/kVB**, because we *love* vbytes for fee-sensitive Bitcoin applications. Params are `bitcoind`-style.

This is based on #3480 as it uses its higher-level JSONRPC helpers for plugins.

### TODO
- [x] bcli-specific functional tests
- ~~interface documentation, for helping Bitcoin L1 developers to develop a plugin to serve as a C-lightning Bitcoin backend :)~~

### Next
- Add the possibility for the Bitcoin plugin to send notifications to `lightningd` (like "got a new block")
- Add a Bitcoin backend plugin for [insert Bitcoin data source]

darosior commented:
A minimal plugin which uses an esplora instance instead of `bitcoind` to serve Bitcoin data to C-lightning can be found here https://github.com/lightningd/plugins/pull/89.

ZmnSCPxj commented:
Spurious failure:

>      lightningd/lightningd: --bitcoin-rpcuser=rpcuser: unrecognized option

https://travis-ci.org/ElementsProject/lightning/jobs/647097656#L1083

That seems a strange error to report.

darosior commented:
> That seems a strange error to report

This is just that I remove the default plugin in one of the tests, to test the Bitcoin plugin registration behaviour.
But I'll give it a look anyway.

darosior commented:
Still one unrelated spurious Travis error, and two jobs timeout..

darosior commented:
I've removed the doc from the TODOs after discussion with @cdecker on IRC so that it doesn't lock up with that interface too early. I had already written it so if reviewers find it useful, you can find it [here](https://gist.github.com/darosior/92287246d26b5accb34d3381577126f0).

rustyrussell commented on lightningd/bitcoind.c:
```diff
@@ -38,6 +38,43 @@ static const char *methods[] = {"getchaininfo", "getrawblockbyheight",
                                 "sendrawtransaction", "gettxout",
                                 "getfeerate"};
 
+static void plugin_config_cb(const char *buffer,
```
"We need our Bitcoin backend to be initialized, but the plugins have not yet been
started at this point."

(GH does not allow comments on commit msgs).

This seems like a can of worms.

OTOH, reordering to create the rpc before topology is ready is also problematic: there could innocently be an RPC call requiring topology information (plugin init, or anyone else) and the only sane semantic is to block at that point until we have it.

Can we create the rpc file, but not service it until we have topology?  That puts a burden on any `bitcoind` plugin not to deadlock, but it's simple?

darosior replied:
Ok, I had some edge cases when I tried but now it's working fine.

I'm still not sure about requiring the RPC socket to be initialized in `libplugin`, especially for plugins ""under"" `lightningd`.

darosior replied:
Hmmm I'll have to make the plugins initialization synchronous then

darosior replied:
> but not service it until we have topology

Ok I missed this part.

rustyrussell commented on lightningd/bitcoind.c:
```diff
@@ -571,6 +544,84 @@ static void bitcoin_plugin_error(struct bitcoind *bitcoind, const char *buf,
+				     "sendrawtx",
+				     "bad 'success' field");
+
+	errtok = json_get_member(buf, resulttok, "errmsg");
```
Prefer not to require errtok unless !success.

rustyrussell commented on lightningd/bitcoind.c:
```diff
@@ -784,89 +784,88 @@ void bitcoind_getchaininfo_(struct bitcoind *bitcoind,
 }
 
-struct get_output {
+/* `gettxout`
```
Can we call this `getutxout` to emphasize that it's only UTXOs?

darosior replied:
Done.

darosior commented on lightningd/bitcoind.c:
```diff
@@ -1149,6 +1152,7 @@ struct bitcoind *new_bitcoind(const tal_t *ctx,
 {
 	struct bitcoind *bitcoind = tal(ctx, struct bitcoind);
 
+	strmap_init(&bitcoind->pluginsmap);
```
We want to be able to make distinct plugins register different commands.

darosior commented on plugins/libplugin.h:
```diff
@@ -212,7 +212,8 @@ struct command_result *timer_complete(struct plugin *p);
 				  struct timerel t,
-				  struct command_result *(*cb)(struct plugin *p));
+				  void (*cb)(void *cb_arg),
+				  void *cb_arg);
```
I'm happy to do it, but I wasnt sure which one to pick since the two callers don't use the same type ?

darosior replied:
> make sure the cb_arg matches the last argument of the cb function

Ah ok ! I'll put up a PR :-)

THanks for the issue

darosior commented:
> Only one significant change req, really: split rpc-file creation from rpc servicing, and do creation before starting plugins.

Ok so I tempted but there is a circular dependency which causes a deadlock.
If we want the JSONRPC interface to be set up before topology but not serve requests until topology is setup, we can simply:
```patch
diff --git a/lightningd/jsonrpc.c b/lightningd/jsonrpc.c
index d2f896904..15a2df66a 100644
--- a/lightningd/jsonrpc.c
+++ b/lightningd/jsonrpc.c
@@ -984,6 +984,13 @@ static struct io_plan *jcon_connected(struct io_conn *conn,
 static struct io_plan *incoming_jcon_connected(struct io_conn *conn,
                                               struct lightningd *ld)
 {
+       /* The topology is required to be ready by most of the JSONRPC calls,
+        * even the simplest ones (`getinfo` queries the tip). However we
+        * want to serve the RPC socket before we setup the topology (mostly
+        * for plugins), so accept incoming requests but don't treat them yet. */
+       if (!ld->topology->ready)
+               return io_wait(conn, ld->jsonrpc, incoming_jcon_connected, ld);
+
        /* Lifetime of JSON conn is limited to fd connect time. */
        return jcon_connected(notleak(conn), ld);
 }
diff --git a/lightningd/lightningd.c b/lightningd/lightningd.c
index 0874bee54..594383fff 100644
--- a/lightningd/lightningd.c
+++ b/lightningd/lightningd.c
@@ -855,6 +855,11 @@ int main(int argc, char *argv[])
        setup_topology(ld->topology, ld->timers,
                       min_blockheight, max_blockheight);
 
+       /* Now that the topology is set up, we can treat incoming JSONRPC
+        * requests ! */
+       assert(ld->topology->ready);
+       io_wake(ld->jsonrpc);
+
```
But we need to start plugins in order to setup the topology.... And plugins use `libplugin` which requires the JSONRPC to answer a request (`listconfigs` in order to set the `allow_deprecated_apis` global) before responding to `init`.

So either we get rid of this call in `handle_init`, which I believe is not acceptable, or we setup the plugins needed to setup the topology before the RPC socket (and other current plugins).

I think the right thing would be to make a distinction between plugins which rely upon `lightningd` (and thus need the RPC socket, are setup after the topology, etc..), and plugins which `lightningd` relies upon.
That said, with this PR the only plugin(s) `lightningd` relies on are `bitcoind`-ones, so it's special-cased into `lightningd/bitcoind`(and this is cleaner than all my attempts to create the socket before starting these plugins).

darosior commented:
Rebased, renamed `gettxout` to `getutxout`, removed the requirement for `errmsg` to be present if `sendrawtransaction` succeeds, and generalized `find_plugin_for_command()`.

The doc is ready but I wanted @cdecker feedback before as he had a reserve about documenting too early.

rustyrussell commented:
Yeah, OK, let's do this for now then.  Nasty stuff though :(

rustyrussell commented:
Trivial rebase.  Still needs a Changelog entry though!

darosior commented:
Rebased and changeloged

cdecker commented on lightningd/bitcoind.c:
```diff
@@ -420,59 +119,74 @@ struct estimatefee {
+	json_add_string(req->stream, "mode", call->estmode[call->i]);
+	jsonrpc_request_end(req);
+	plugin_request_send(strmap_get(&bitcoind->pluginsmap,
+				       "getfeerate"), req);
```
Would be nice if we could bundle these and send all the block/mode combinations to the plugin at once. That'd allow us to have a single all-or-nothing handler and not having to loop in `lightningd`.

darosior replied:
Indeed. As a cleanup or before the release ?

rustyrussell replied:
Post.  Too late for release now.

darosior replied:
Opened https://github.com/ElementsProject/lightning/issues/3508 to track this.

cdecker commented on lightningd/bitcoind.c:
```diff
@@ -80,6 +80,9 @@ void bitcoind_check_commands(struct bitcoind *bitcoind)
 		if (p == NULL) {
+			/* For testing .. */
+			log_debug(bitcoind->ld->log, "Missing a Bitcoin plugin"
+						     " command");
```
Should probably include the method name that can't be resolved.

darosior replied:
This is just used in functional tests, the `fatal` below includes it

cdecker commented on tests/test_plugin.py:
```diff
@@ -932,3 +932,84 @@ def test_hook_chaining(node_factory):
+            assert l1.daemon.is_in_log(msg)
+        else:
+            raise Exception("We registered all commands but couldn't start!")
+    else:
```
This nesting of `try`-`except` seems strange to me.

darosior replied:
I tried to test for errors, if you have a more elegant idea ? I could still put them one after the other but it's the same

cdecker commented on tests/test_plugin.py:
```diff
@@ -932,3 +932,84 @@ def test_hook_chaining(node_factory):
+                                                   "mode": "CONSERVATIVE"})
+
+    resp = l1.rpc.call("getchaininfo")
+    assert resp["chain"] == "regtest"
```
This needs mocking as well since we may be running on `liquid-regtest`.

darosior replied:
Ah forgot this one thanks!

darosior commented on lightningd/bitcoind.c:
```diff
@@ -556,34 +556,211 @@ void bitcoind_getrawblock_(struct bitcoind *bitcoind,
+	if (json_tok_is_null(buf, blockhashtok)) {
+		db_begin_transaction(call->bitcoind->ld->wallet->db);
+		call->cb(call->bitcoind, NULL, NULL, call->cb_arg);
+		db_commit_transaction(call->bitcoind->ld->wallet->db);
```
I don't see why it should: `lightningd/io_loop_with_timers` only begins a db_transaction is a timer expires (out of the `io_loop`).

Before this `bcli_finished` would make a db transaction for each callback https://github.com/ElementsProject/lightning/blob/86c28b2272b65fb0926b919ea0c82327ba3051de/lightningd/bitcoind.c#L224-L226

darosior commented:
Spurious Travis error

cdecker commented:
ACK https://github.com/ElementsProject/lightning/commit/93596e36cda393aea7df9e35a240a3f34c563e7f

rustyrussell commented:
Apparently not :(  Mind you, GH doesn't give anything for that 'force-push' link either...

rustyrussell commented:
Squashed fixup:

Ack f7b48ed0879c1943a7afd7076cd9e753472895e0

cdecker commented:
Interestingly [the web frontend](http://46.101.246.115:8888/ElementsProject/lightning/pull/3488) seems to recognize them as identical (green background for the comparison between the last two versions). I'll look into it
